Single package fixpoint step 2: introduce `mkPackage`

roberth opened the pull request:
## Description of changes

This provides an alternative to `mkDerivation` - a different interface which solves a significant number of problems.

- Proposal: #273815

- Includes: #158781 

- Follows #295378

Features and problems solved:
- Remove the buggy separation between
- Decouple [derivations](https://nix.dev/manual/nix/2.25/glossary#gloss-derivation) from [package attrsets](https://nix.dev/manual/nix/2.25/glossary#package-attribute-set) (specifically `mkPackage` will be usable with alternatives to `layers.derivation`)
  - Facilitate multi-derivation packages, e.g. separate `doc` output to improve bootstrapping and avoid some rebuilds
  - Facilitate [RFC 92 dynamic derivations](https://github.com/NixOS/rfcs/blob/master/rfcs/0092-plan-dynamism.md) packages
  - https://github.com/NixOS/nix/issues/6507
- Clean up the interface of packages
  - Make attributes more relevant
  - Incentivize more explicit interfaces, more knowable public interface
  - Successor of https://github.com/NixOS/nixpkgs/pull/217243
- Ability to have overridable package variables that do not affect the derivation and aren't in `passthru`
  - Makes derivation attributes intentional
- Modularity
  - Language integrations in `layers` can be combined, e.g. python + javascript in a single derivation


TODO
- [ ] Port this [fileset layer](https://github.com/NixOS/nix/pull/10995/files#diff-1904adb2564cf5e20a791943a2f426eb6c9525ef341476e729c9325797e8c96bR16-R37). It will be nicer, not having to "unset" things for `derivation`.
- [ ] rename `setup`, i.e. where known stdenv/setup/mkDerivation arguments go
- [ ] rebase the old `mkDerivation` interface on top of this?


## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
lib: Add encapsulate

Creates objects that have overlay-based private attrs in their closure.

(cherry picked from commit 0067cf4fc2ff48a4cf59e961c1917e9bf2faec36)

roberth committed:
Add mkPackageWithDeps

This shows the `deps` pattern, merging the following into a
single fixpoint, from https://github.com/NixOS/nixpkgs/issues/273815

 - package function args (`this.deps`)
 - mkDerivation args (`this.setup`)
 - derivation args (`this.drvAttrs`)
 - package attributes (`this.public`)

roberth committed:
mkPackageWithDeps: Allow non-pkgs deps

Demo:

```diff
diff --git a/pkgs/by-name/he/hello/package.nix b/pkgs/by-name/he/hello/package.nix
index 0cf0ae493b72..69921e14ffc6 100644
--- a/pkgs/by-name/he/hello/package.nix
+++ b/pkgs/by-name/he/hello/package.nix
@@ -1,9 +1,13 @@
-{ mkPackageWithDeps, layers }: mkPackageWithDeps ({ stdenv, fetchurl, testers, lib, callPackage, nixos }: [
+{ mkPackageWithDeps, layers }: mkPackageWithDeps ({ stdenv, fetchurl, testers, lib, callPackage, nixos, haskellPackages, omnidoc }: [
   (layers.derivation { inherit stdenv; })
   (this: old: {
     name = "hello";
     version = "2.12.1";

+    deps = old.deps // {
+      omnidoc = haskellPackages.pandoc;
+    };
+
     setup = old.setup or {} // {
       src = fetchurl {
         url = "mirror://gnu/hello/hello-${this.version}.tar.gz";
@@ -24,6 +28,10 @@
       run = callPackage ./test.nix { hello = this.public; };
     };

+    public = old.public // {  # passthru, just to show that it works
+      inherit omnidoc;
+    };
+
     meta = with lib; {
       description = "A program that produces a familiar, friendly greeting";
       longDescription = ''
```

Ericson2314 commented:
So continuing from https://github.com/NixOS/nix/pull/10610#issuecomment-2080102957, my thought is similar to the talk of NixOS module system modularity. Just as we'd like multiple PostgreSQL "instances", for example, I think we should be very careful to destinguish a package in isolation, which does not care how it is used (e.g. choosing its deps) --- it is function, from an instantiate package in a package set, which is not a function.

roberth commented:
> I think we should be very careful to destinguish a package in isolation, which does not care how it is used (e.g. choosing its deps)

I agree. For this it is important that the package presents itself as a function with a stable interface.

Suppose you're packaging `app`, which is a function of `{ stdenv, postgresql }`, but it doesn't work with the default postgresql version in Nixpkgs.
In the current architecture you have two options:
1. make it a function of `{ stdenv, postgresql_12 }`
2. set a default value, `callPackage ../by-name/ap/app/package.nix { postgresql = postgresql_14; }`.

In this situation, (1) is wrong because it chooses its deps, and it's not a stable function this way, because callers need to be aware of the current default version.
However (2) is not desirable; contributors don't want to touch `all-packages.nix`, if they even believe they're allowed to.

With this implementation of the single fixpoint idea, they can preserve the existing function interface, and set a default in the same file - without touching `all-packages.nix`:

`package.nix`:
```nix
{ mkPackageWithDeps, layers }:
mkPackageWithDeps ({ pkgs, stdenv, postgresql, ... }: [
  (layers.derivation { inherit stdenv; })
  (this: old: {
    name = "app";
    deps = old.deps // {
      postgresql = pkgs.postgresql_14;
    };
    setup.buildInputs = [ postgresql ];
  }
])
```

`deps.postgresql` has a default, but its value can still be set with `callPackage ./package.nix { postgresql = ...; }` or `.override`.
It is still a proper function that way, and its signature is stable.

Ericson2314 commented:
But what happens if I try to instantiate this package within a package set that doesn't have a `postgresql_14` attribute?

I would prefer something like a separate `package.nix` and `deps.nix`, or a `package.nix` with

```nix
{
   function = ...;
   dep-defaults = ...;
}
```

for a bit more of a by-construction guarantee the package function stands alone without the defaults.

roberth commented on pkgs/by-name/he/hello/package.nix:
```diff
@@ -1,46 +1,41 @@
-, testers
-, hello
-}:
+{ mkPackageWithDeps, layers }: mkPackageWithDeps ({ stdenv, fetchurl, testers, lib, callPackage, nixos }: [
```
Consider adding `layers` as well as a function `overrideLayers` to the call here, such that they don't come from the normal fixpoint. Otherwise it's not possible to override that `layers`, which someone will inevitably need. 

It also makes the separation of concerns a bit nicer:
```nix
# 1 parameter to declare that we'll be doing it the layered way
{ mkPackageWithDeps }: mkPackageWithDeps ({
  # all the rest goes here
  layers, ...
}: [
  # ...
])
```

That changes the fixpoint "state" from just overlays to `{ overlays = <composition of functions>; layers = <attrset of functions brought into scope by mkPackageWithDeps>; }`, but that seems manageable and not something users have to be aware of, as long as it works correctly. It's still just a single fixpoint, so it seems feasible to get it right.

roberth commented:
> * a `package.nix` with
>       ```nix
>       {
>          function = ...;
>          dep-defaults = ...;
>       }
>       ```

This could be implemented as syntax sugar. The `package.nix` file can choose how to construct itself:

```nix
{ mkPackageWithDefaults }: mkPackageWithDefaults {
  function = ...;
  dep-defaults = ...;
}
```

If we don't implement it as syntax sugar, the "obvious" code would not have a single fixpoint, which would instantly make overriding incoherent. Make it syntax sugar, and we avoid that problem.

Ericson2314 commented:
> The `package.nix` file can choose how to construct itself

Ah, but this is just what I am taking issue with, I want to *force* packages to *not* mix together the package function and the default argument choosing. I agree making people edit `all-packages.nix` is bad, but I think the separation between choosing deps and declaring dep parameters is very good.

If we think packages are "straight away" put in the big fixed point, this can all seem academic, but if we think we'll first build a library of package functions, and *then* build the big fixed point from that, the distinction gets more practice.

The "library of package functions" (recipe book) is very important to me as the only thing which is truly modular.

roberth commented:
> Ah, but this is just what I am taking issue with, I want to _force_ packages to _not_ mix together the package function and the default argument choosing.

Sure, but then you're not compatible with `by-name/*/*/package.nix`. I'd be ok with enforcing something, such as by loading a different file that only has the expression you suggested, but that seems like a constraint that's easy to add later on.

> if we think we'll first build a library of package functions

That sounds a lot like what was done in the past and what has got us into the current situation, unless we have the discipline not to add unnecessary fixpoints in the process, as you suggest.

I am a bit skeptical though:
- I doubt that a huge bag of functions is a pleasant to use intermediate state of affairs.
- Without any conventions emanating from the single fixpoint, we'd probably end up with functions that wrap other functions, such as the language-specific builders. Even if those don't have their own fixpoint logic, they're going to need a degree of rewriting to fit the those conventions, once we do establish them in the end.

Ericson2314 commented:
> That sounds a lot like what was done in the past and what has got us into the current situation, unless we have the discipline not to add unnecessary fixpoints in the process, as you suggest.

Yeah I want to do one round of importing all the functions and calling nothing, maximally flexible. No fixed points, no overriding, non of that. Just library of recipes, everything is in "fully functored form" as the OCaml people would say.

And then I want to do another round of one big fixed point, putting everything together.

> Without any conventions emanating from the single fixpoint, we'd probably end up with functions that wrap other functions, such as the language-specific builders.

You are contrasting that the layers thing you have? I haven't thought about that so much but I think there is no conflict.

roberth commented:
@Ericson2314 I think we agree on the overall strategy.

Atemu commented:
> Yeah I want to do one round of importing all the functions and calling nothing, maximally flexible. No fixed points, no overriding, non of that.

I particularly like this idea because it could make it possible to "override" elements of the package set that are not drvs or attrsets in their "final form" (where a `.override` attr could be added) such as builders or wrappers.

purepani commented:
@roberth This looks really cool! I was wondering if you had mocked up a use of this type of function in a similar manner to dream2nix(since you mentioned it in the issue); I was thinking of trying to do that both as a learning opportunity and to see if I could figure out dynamic derivations as well, but I also didn't want to redo anything you were working on, especially since these things would probably be pretty hard for me given that all of this is pretty new. 

Also wanted to ask @DavHau since you also might be working on something like this, or might have some insight.

roberth commented:
@purepani, that sounds like a lot to do simultaneously. My goal was to show that it can be done, and establish a starting point. Haven't been able to spend much time on it since. I'd be happy to explain things if you want to try it.

nbp commented:
While I definitely agree and I am pleased with the result. I would have one recommendation which is to make it at first such that the existing function can still be used without making any changes, and then progressively replace one by the other.

The idea would be to avoid having any down-time as these kind of changes would likely takes months / years before they are complete, and having something backward compatible would also help with the migration process.

Thus `callPackage` should become a function which create a set with all the inputs and the function the be called; `stdenv.mkDerivation` should be a function which extends the result of `callPackage` with the `drvOut` public attributes.

roberth commented:
@nbp Not sure I understand what you mean by downtime. This PR adds an alternative to `mkDerivation`, which packages can migrate to on an individual basis. The returned / public attributes are largely compatible with the existing packages (except being more intentional / "cleaner", which should be fine, and if not, we could have a layer that adds back a bunch of bloat).



> Thus `callPackage` should become a function which create a set with all the inputs and the function the be called

That sounds rather invasive. `callPackage` is widely used for
 - producing packages
 - performing dependency injection

I don't see how returning functions and arguments helps with those expectations or reducing the number of fixpoints per package. Rather it hints at adding another one.
My very intentional intent behind the change to `lib.customisation.callPackageWith` is to _remove_ `.override` from its responsibilities when the package definition is a single-fixpoint package. That way it can be handled simply and correctly by the package fixpoint.
If we were to make `callPackage` return `result // { inherit origArgs function; }`, we make available exactly the info that allows `.override` to be reconstructed and re-added to the package. That works against the goal of forcing all package-based overrides to go through a single fixpoint. And if we don't go through a single fixpoint, we'll end up recreating those same composition / correctness issues this is supposed to solve.

roberth commented:
I think you're applying a solution you've thought of for S.O.S. here, but prematurely. My understanding of that concept is that it operates at the package _set_ level, unlike this PR.
Once packages consist of a single fixpoint, it will be easy to "open them up" and let them be specified as a single function instead of a file that also ties the knot. I can (vaguely) see how that could then enable simplifications at the package _set_ level.

So I'd be happy to replace `callPackage` at that point, but until then, this has to work with the existing large-scale infrastructure of Nixpkgs, and since RFC 140, that basically also means hooking into `callPackage` somehow. If it wasn't for that, I'd be happy to sidestep `callPackage`, and avoid the `{ mkPackageWithDeps, layers }: mkPackageWithDeps` opening. The purpose of that is just to integrate with Nixpkgs how it works today, selecting a package producing helper just like packages now pick `stdenv.mkDerivation` or a language-specific one. When the single fixpoint approach is adopted, this decision is reduced to a selection of layers instead of a choice of function, and we'll be able to remove all those openings and let the packages just be a function for the single package fixpoint.

roberth commented:
Maybe what you're asking for is a way to skip ahead to that point where we can change all package files to be single-fixpoint style?
I think we could come up with a layer function that basically just calls the existing `mkDerivation`, language-specific functions, etc, so that all package files can migrate mechanically without much of the benefits of actually having a single fixpoint.
When that's done, we remove those openings and something like S.O.S. could `fix` those functions as it pleases.

purepani commented:
> @purepani, that sounds like a lot to do simultaneously. My goal was to show that it can be done, and establish a starting point. Haven't been able to spend much time on it since. I'd be happy to explain things if you want to try it.

I'll definitely let you know if I have any questions once I have a chance to dig into it!

purepani commented:
Having some time to think about this, I'm wondering if it would make sense to have some sort of "typing information" file(like header files in C++), and sort of introduce gradual typing into this.
That's probably beyond the scope of this idea, but I'm only wondering that because it might make the whole merging speed issue faster, if we can, say, write a typing file for a package that defines an interface(as mentioned earlier), of what the outputs of it are(in terms of typing) and the inputs are.
I'm imagining taking, for example, the already existing package.nix file and writing a typing.nix file that looks like this
```
{lib, postgres}:
inputs = {
   postgres=postgress.type
};
outputs={
out=lib.type.out;
fetch-deps=lib.type.fetchdeps;
};
```

Then you could prerun, and cache the typing information, and then use that information to help in any sort of merging process, only recompiling when any header files change.

This feels like something that could be more useful to implement on a language level, but I thought I'd put it out there since it's an extension of what was being talked about earlier.

roberth commented:
> merging speed issue faster

Something like [this](https://gist.github.com/roberth/940dff88ca5f5f95949dc309dbe60a65#file-minimod-nix-L126) is probably half way in terms of performance between the module system and hand-coded merging (as we do with `overrideAttrs` etc).

We could go faster by removing the priorities, or slower by always checking the value. To really go faster though, I think we need a `builtins.zipAttrsWith'` that doesn't perform a Nix function call when there's only a single value.

To bring this closer, we could make a gradual change to `encapsulate`. Instead of
`{ extend = f: encapsulate (extends f layerZero); }` in `encapsulate`, we could do something more general, perhaps turning the hardcoded `lib.extends` into a parameter; `{ extend = x: encapsulate extender (extender layerZero x); }`, where `x` could be `lib.extends` or something that helps with merging. Or `{ extend = x: encapsulate ((fixed.extender or extends) layerZero x); }`, adding no memory cost in the default case where it's just `extends`.
A potential limitation of this approach is that it potentially requires reassembling the typing metadata for every instance, while it would be preferable to only assemble, say, an attrset of merge functions only once per language specific build helper (ie one of the `generic-builder.nix`-es).

Meanwhile @nbp is working on a proposal (in his free time, afaik) to make deep overrides easier to write, which _might_ also run faster in the evaluator just by virtue of being interpreted more directly instead of a bunch of function calls with all the Nix function bells and whistles. Idk, maybe we should get someone from SpiderMonkey on Nix ;)

roberth commented:
I think the next step is to try and convert an existing `generic-builder.nix` to use this, without trying to change the data model - ie preserve all the nasty `removeAttrs` and `//` operations on the recipe function (the "user"-provided function), so that the packaging logic is 1:1 compatible, or at least very close.
- Feasible to actually migrate
- No "improved" data model for users (but a path towards that maybe?)
- Can fix overriding composition complexity and bugs

Steps
1. Extract functions from `generic-builder.nix` that perform the attribute manipulation, not including calls to `.override*` or `lib.customisation`. (as previously suggested by @Ericson2314)
2. Create a new function that calls `lib.encapsulate` and uses those functions to achieve the same wiring.
3. Try instantiating the applicable packages with it, see if anything breaks.
4. (Optional) Create a variation of the function that also includes the layer with `deps` to replace `callPackage` and fix the `override`/`overrideAttrs` interaction.
   This will require package to make a change at the top of the file, e.g.
   `{ mkLangDerivation, foo, bar }: mkLangDerivation { ... }`
   -> `{ mkLangDerivationWithDeps }: mkLangDerivationWithDeps ({ foo, bar }: { ... })`

emilazy commented:
I really hope that this can support an interface like https://github.com/NixOS/nixpkgs/issues/227327. IMO that’s the only way we’ll ever get a truly usable cross‐compilation interface and finally kill off splicing. I think that moving away from the dependencies‐as‐function‐arguments style will both reduce duplication and make overrides more usable.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/is-it-possible-for-us-to-remove-builtins-functionargs/51960/4

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/community-team-updates/56458/11

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/whats-the-status-of-the-single-package-fixpoint-proposal/58692/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/whats-the-status-of-the-single-package-fixpoint-proposal/58692/2

roberth committed:
Make layers overridable

This also makes the old `mkPackageWithDeps` style the default, and
improves error reporting a bit.

roberth committed:
callPackage: Special treatment for mkPackage

The solution inside makeOverridable was fragile, and did not catch
the potential issue that the new code reports.

It's still not super pretty, but architecturally it won't get better
than this because `by-name/**/package.nix` is tied to `callPackage`.

Alternative solution: require a different file name, so that
`callPackage` can be avoided.

roberth committed:
layers.buildNpmPackage: init

roberth committed:
netlify-cli: Use mkPackage

This allows the all-packages.nix entry to be removed
thanks to pkgs/by-name 🎉

roberth committed:
mkPackage: Expose fixpoint root as .internals

Giving no access at all would be really annoying for debugging.

Unlike `mkDerivation`, which makes no distinction between intentional
and internal package attributes, this gives a signal that when you're
using it, you are more likely to have to change your code at some point.
(besides keeping the public package attrs _relevant_)

roberth committed:
callPackage: Add comment about callPackage/mkPackage usage

WeetHet commented:
@roberth what is the plan for splicing? There are similar efforts to this in #371613 which seem to use pkgsOnHost/pkgsOnBuild

roberth committed:
fixup! netlify-cli: Use mkPackage

roberth commented:
> plan for splicing?

Nothing special for now. `mkPackage` inherits the `callPackage` behavior, so that works as it did before.
You can see it in operation with the `hello` package, and I've confirmed that a cross build of it works as expected.
`mkPackage` could be a good place to implement changes to splicing or cross compilation more generally, but I wouldn't expand the scope to encompass all of https://github.com/NixOS/nixpkgs/issues/273815 in one PR.

roberth committed:
lib.customisation.hydraJob: system is optional

The real _derivation system_ is best retrieved from `drvPath`.
Putting this in an attribute with this simple name is ambiguous,
because it can refer to either the `hostPlatform.system` or
`buildPlatform.system`, which are unequal in cross compilation.
The _system_ attribute tends to refer to the _derivation system_
which is the `buildPlatform.system`, but this is an implementation
detail of the package internals, and it is not relevant to any
in-language consumers of the package, especially considering that
knowledge about one `system` does not imply anything about the many
possible `system`s that can be in the derivation's build closure,
which tends to be equally relevant for the process of realisation.

roberth committed:
hello.src: Restore

roberth commented on lib/fixed-points.nix:
```diff
@@ -451,6 +451,96 @@ rec {
+  encapsulate =
+    layerZero:
+    let
+      fixed = layerZero ({ extend = f: encapsulate (extends f layerZero); } // fixed);
```
This forms the new fixpoint.

roberth commented on lib/customisation.nix:
```diff
@@ -301,7 +301,26 @@ rec {
+            ])
+          ''
+        else
+          f allArgs
```
This avoids adding the second fixpoint (`callPackage`/`makeOverridable`/`.override`).

I should say that this is an overly broad interpretation of fixpoint, but a recursion is definitely there, so I'll just use the term. (even if `result` isn't passed back to `makeOverridable`'s `f`, the overriding dynamics are equally problematic)

roberth reviewed (commented):
Just pointing out how the fixed points change; see review comments

+1 fixpoints: `encapsulate`
-1 fixpoints: `callPackage`
-1 fixpoints: `mkDerivation`

for a total change of -1 accidentally individually overridable things

roberth committed:
test: Accept streamNixShellImage reliance on internals for now

roberth commented:
A single fixpoint is still one too many in some cases.

For instance, a package like Nix now consists of multiple derivations for the various components. It holds the middle between a package and a package set. In order to facilitate things like overriding the shared `src`, we need to deal with multiple derivations, each of which is also exposed as a package.
Enabling this brings us somewhat closer to @nbp's [Simple Overriding Strategy](https://github.com/NixOS/rfcs/pull/3), which could be framed as "Single nixpkgs fixpoint", which I don't know if that's possible due to bootstrapping, but is at least like a point on the horizon so to speak.
What I believe this would entail concretely for this "single _package_ fixpoint" PR is that it may be worth looking into whether it could provide a bit of infrastructure to support not tying the fixpoint knot but managing just the function instead. However, it seems likely that this exploration would balloon the scope too much, and it would better be done in a follow-up with more focus on this package set aspect. Walk before you can run, that kind of thing.
